packages/react-form/src/hooks/form.ts

```diff
@@ -50,36 +60,56 @@ import {useSubmit} from './submit';
  * @param fields - A dictionary of `Field` objects, dictionaries of `Field` objects, and lists of dictionaries of `Field` objects. Generally, you'll want these to be generated by the other hooks in this package, either `useField` or `useList`. This will be returned back out as the `fields` property of the return value.
  *
  * @param onSubmit - An async function to handle submission of the form. If this function returns an object of `{status: 'fail', error: FormError[]}` then the submission is considered a failure. Otherwise, it should return an object with `{status: 'success'}` and the submission will be considered a success. `useForm` will also call all client-side validation methods for the fields passed to it. The `onSubmit` handler will not be called if client validations fails.
+ * @param lists - Pass in list objects like dynamic list into the form
+ * @param makeCleanAfterSubmit
  * @returns An object representing the current state of the form, with imperative methods to reset, submit, validate, and clean. Generally, the returned properties correspond 1:1 with the specific hook/utility for their functionality.
  *
  * @remarks
  * **Building your own:** Internally, `useForm` is a convenience wrapper over `useDirty`, `useReset`, and `useSubmit`. If you only need some of its functionality, consider building a custom hook combining a subset of them.
  * **Subforms:** You can have multiple `useForm`s wrapping different subsets of a group of fields. Using this you can submit subsections of the form independently and have all the error and dirty tracking logic "just work" together.
  */
+
 export function useForm<T extends FieldBag>({
   fields,
   onSubmit,
+  dynamicLists,
   makeCleanAfterSubmit = false,
 }: {
   fields: T;
   onSubmit?: SubmitHandler<FormMapping<T, 'value'>>;
   makeCleanAfterSubmit?: boolean;
+  dynamicLists?: DynamicListBag;
 }): Form<T> {
-  const dirty = useDirty(fields);
-  const basicReset = useReset(fields);
+  const fieldsWithLists = useMemo(() => {
+    const fieldsWithList = {...fields};
+    if (dynamicLists) {
+      Object.entries(dynamicLists).forEach(([key, value]) => {
+        (fieldsWithList as any)[key] = value.fields;
+      });
+    }
+    return fieldsWithList;
+  }, [dynamicLists, fields]);
+
+  const dirty = useDirty(fieldsWithLists);
+  const basicReset = useReset(fieldsWithLists);
+
+  const dynamicListDirty = useDynamicListDirty(dynamicLists);
+  const dynamicListReset = useDynamicListReset(dynamicLists);
+
   const {submit, submitting, errors, setErrors} = useSubmit(
     onSubmit,
-    fields,
+    fieldsWithLists,
     makeCleanAfterSubmit,
   );
 
   const reset = useCallback(() => {
     setErrors([]);
     basicReset();
-  }, [basicReset, setErrors]);
+    dynamicListReset();
+  }, [basicReset, dynamicListReset, setErrors]);
 
-  const fieldsRef = useRef(fields);
-  fieldsRef.current = fields;
+  const fieldsRef = useRef(fieldsWithLists);
+  fieldsRef.current = fieldsWithLists;
```

Comment:
I know this was already here, but a note on the pattern: 

This technically works but may stop working in future react updates. We recently removed similar patterns in [other hooks in quilt](https://github.com/Shopify/quilt/pull/1813) and are considering adding a [lint rule](https://github.com/Shopify/web-configs/issues/229). It might be best to find an alternative way of doing this as a result.

Actually, looking at it closer, since we update it literally every render, why can't we just use the value directly?

Reply:
Ahh yes I think we can get rid of the ref since we update it, Thanks! Will try that